src/progress.rs

```diff
@@ -0,0 +1,308 @@
+    },
+    thread::JoinHandle,
+    time::{Duration, Instant},
+};
+
+/* -------------------------------------------------------------------------- */
+
+pub const ATOMIC_ORDERING: Ordering = Ordering::Relaxed;
+
+// small wrappers for atomic number to reduce overhead
+pub trait ThreadSyncTrait<T> {
+    fn set(&self, val: T);
+    fn get(&self) -> T;
+}
+
+pub trait ThreadSyncMathTrait<T> {
+    fn add(&self, val: T);
+}
+
+macro_rules! create_atomic_wrapper {
+    ($ident: ident, $atomic_type: ty, $type: ty, $ordering: ident) => {
+        #[derive(Default)]
+        pub struct $ident {
+            inner: $atomic_type,
+        }
+
+        impl ThreadSyncTrait<$type> for $ident {
+            fn set(&self, val: $type) {
+                self.inner.store(val, $ordering)
+            }
+
+            fn get(&self) -> $type {
+                self.inner.load($ordering)
+            }
+        }
+    };
+
+    ($ident: ident, $atomic_type: ty, $type: ty, $ordering: ident + add) => {
+        create_atomic_wrapper!($ident, $atomic_type, $type, $ordering);
+
+        impl ThreadSyncMathTrait<$type> for $ident {
+            fn add(&self, val: $type) {
+                self.inner.fetch_add(val, $ordering);
+            }
+        }
+    };
+}
+
+create_atomic_wrapper!(AtomicU64Wrapper, AtomicU64, u64, ATOMIC_ORDERING + add);
+create_atomic_wrapper!(AtomicU8Wrapper, AtomicU8, u8, ATOMIC_ORDERING + add);
+
+#[derive(Default)]
+pub struct ThreadStringWrapper {
+    inner: RwLock<String>,
+}
+
+impl ThreadSyncTrait<String> for ThreadStringWrapper {
+    fn set(&self, val: String) {
+        *self.inner.write().unwrap() = val;
+    }
+
+    fn get(&self) -> String {
+        (*self.inner.read().unwrap()).clone()
+    }
+}
+
+/* -------------------------------------------------------------------------- */
+
+// creating an enum this way allows to have simpler syntax compared to a Mutex or a RwLock
+#[allow(non_snake_case)]
+pub mod Operation {
+    pub const INDEXING: u8 = 0;
+    pub const PREPARING: u8 = 1;
+}
+
+#[derive(Default)]
+pub struct PAtomicInfo {
+    pub file_number: AtomicU64Wrapper,
+    pub files_skipped: AtomicU64Wrapper,
+    pub directories_skipped: AtomicU64Wrapper,
+    pub total_file_size: TotalSize,
+    pub state: AtomicU8Wrapper,
+    pub current_path: ThreadStringWrapper,
+}
+
+impl PAtomicInfo {
+    fn new(c: &PConfig) -> Self {
+        Self {
+            total_file_size: TotalSize::new(c),
+            ..Default::default()
+        }
+    }
+}
+
+/* -------------------------------------------------------------------------- */
+
+#[derive(Default)]
+pub struct TotalSize {
+    use_iso: bool,
+    inner: AtomicU64Wrapper,
+}
+
+impl TotalSize {
+    fn new(c: &PConfig) -> Self {
+        Self {
+            use_iso: c.use_iso,
+            ..Default::default()
+        }
+    }
+
+    fn format_size(&self) -> String {
+        let inner = self.inner.get();
+        let number_len = (inner as f32).log10().floor() as u32;
```

Comment:
could probably use  human_readable_number in display file.

Reply:
I've implemented this, but I think it's better if there is a space between the suffix and the text. What do you think about this ?

Reply:
I think I prefer it without the space. That's what 'du' does. Plus for the tree view I need all the horizontal screen real estate I can get.